diff --git a/drivers/hid/hid-nanosic-wn8030.c b/drivers/hid/hid-nanosic-wn8030.c
index 37b353ede5f4..31d3427631c0 100644
--- a/drivers/hid/hid-nanosic-wn8030.c
+++ b/drivers/hid/hid-nanosic-wn8030.c
@@ -14,6 +14,7 @@
 #include <linux/module.h>
 #include <linux/of.h>
 #include <linux/regmap.h>
+#include <linux/workqueue.h>
 
 #define WN8030_HEADER_ADDR 0x7FC0
 #define WN8030_CODE_ADDR 0x8000
@@ -39,6 +40,7 @@ struct nanosic_wn8030 {
 
 	struct regmap *regmap;
 	struct mutex conn_mutex;
+	struct delayed_work wake_worker;
 
 	bool suspended;
 	bool keyboard_attached;
@@ -360,6 +362,64 @@ static int nanosic_wn8030_set_kb_power(struct nanosic_wn8030 *nanosic, bool enab
 	return regmap_bulk_write(nanosic->regmap, 0x5c, buf, sizeof(buf));
 }
 
+static int nanosic_wn8030_xm_auth_init(struct nanosic_wn8030 *nanosic)
+{
+	u8 buf[XM_WN8030_I2C_WRITE] = { 0x32, 0x00, 0x4F, 0x31,
+					0x80, 0x38, 0x31, 0x06,
+					0x4d, 0x49, 0x41, 0x55, 0x54, 0x48,
+					0x00 };
+
+	buf[14] = nanosic_wn8030_checksum8(&buf[2], 13);
+
+	return regmap_bulk_write(nanosic->regmap, 0x5c, buf, sizeof(buf));
+}
+
+static int nanosic_wn8030_xm_auth_s3t1(struct nanosic_wn8030 *nanosic, u8 *key_meta, u8 *challenge)
+{
+	u8 buf[XM_WN8030_I2C_WRITE] = { 0x32, 0x00, 0x4f, 0x31,
+					0x80, 0x38, 0x32, 0x14,
+					0x00, 0x00, 0x00, 0x00,
+					0x00, 0x00, 0x00, 0x00,
+					0x00, 0x00, 0x00, 0x00,
+					0x00, 0x00, 0x00, 0x00,
+					0x00, 0x00, 0x00, 0x00,
+					0x00 };
+
+	if (key_meta) {
+		memcpy(buf + 8, key_meta, 4);
+	}
+	if (challenge) {
+		memcpy(buf + 12, challenge, 16);
+		buf[28] = nanosic_wn8030_checksum8(&buf[2], 27);
+	}
+	else {
+		buf[12] = nanosic_wn8030_checksum8(&buf[2], 11);
+	}
+
+	return regmap_bulk_write(nanosic->regmap, 0x5c, buf, sizeof(buf));
+}
+
+static int nanosic_wn8030_xm_auth_s5t1(struct nanosic_wn8030 *nanosic, u8 *token)
+{
+	u8 buf[XM_WN8030_I2C_WRITE] = { 0x32, 0x00, 0x4f, 0x31,
+					0x80, 0x38, 0x33, 0x10,
+					0x00, 0x00, 0x00, 0x00,
+					0x00, 0x00, 0x00, 0x00,
+					0x00, 0x00, 0x00, 0x00,
+					0x00, 0x00, 0x00, 0x00,
+					0x00, 0x00, 0x00, 0x00 };
+
+	if (token) {
+		memcpy(buf + 8, token, 16);
+		buf[24] = nanosic_wn8030_checksum8(&buf[2], 26);
+	}
+	else {
+		buf[8] = nanosic_wn8030_checksum8(&buf[2], 10);
+	}
+
+	return regmap_bulk_write(nanosic->regmap, 0x5c, buf, sizeof(buf));
+}
+
 static int nanosic_wn8030_output_report(struct hid_device *hid, u8 *buf, size_t count)
 {
 	struct nanosic_wn8030 *nanosic = hid->driver_data;
@@ -456,16 +516,48 @@ static void nanosic_wn8030_handle_vendor(struct nanosic_wn8030 *nanosic, u8 *buf
 			nanosic_wn8030_add_keyboard_hid(nanosic);
 			nanosic_wn8030_add_touchpad_hid(nanosic);
 			nanosic->keyboard_attached = true;
+			schedule_delayed_work(&nanosic->wake_worker, msecs_to_jiffies(12000));
 		}
 		else if (((buf[12] & 0x3) == 0x0) && nanosic->keyboard_attached) {
-			if (nanosic->hid_keyboard)
+			cancel_delayed_work_sync(&nanosic->wake_worker);
+			if (nanosic->hid_keyboard) {
 				hid_destroy_device(nanosic->hid_keyboard);
-			if (nanosic->hid_touchpad)
+				nanosic->hid_keyboard = NULL;
+			}
+			if (nanosic->hid_touchpad) {
 				hid_destroy_device(nanosic->hid_touchpad);
+				nanosic->hid_touchpad = NULL;
+			}
 			nanosic->keyboard_attached = false;
 		}
 		mutex_unlock(&nanosic->conn_mutex);
 	}
+	/* WN8012(KB) -> HOST, kb auth */
+	// 0x0/0x1 - init auth, 0x64 - repeat request auth
+	else if (buf[5] == 0x38 && buf[6] == 0x80 && buf[7] == 0x24 && buf[9] == 0x64) {
+		nanosic_wn8030_xm_auth_init(nanosic);
+	}
+	/* WN8012(KB) -> HOST, kb auth init */
+	else if (buf[5] == 0x38 && buf[6] == 0x80 && buf[7] == 0x31) {
+		u8 key_meta[4] = {0};
+		u8 challenge[16] = { 0xbe, 0xaf, 0xa1, 0x3b,
+				     0x7c, 0x64, 0x8b, 0xc7,
+				     0xc0, 0x14, 0x85, 0x8c,
+				     0x8b, 0xb2, 0x85, 0x88 };
+		//Select offline key
+		if ((buf[27] & 0x40) != 0x40)
+			memcpy(key_meta, buf + 27, 4);
+		else
+			memcpy(key_meta, buf + 31, 4);
+
+		nanosic_wn8030_xm_auth_s3t1(nanosic, key_meta, challenge);
+	}
+	/* WN8012(KB) -> HOST, kb auth s3t1 */
+	else if (buf[5] == 0x38 && buf[6] == 0x80 && buf[7] == 0x32 && buf[8] == 16) {
+		//u8 token[16] = {0};
+		//nanosic_wn8030_xm_auth_s5t1(nanosic, token);
+		printk("Can be ignored?\n");
+	}
 }
 
 static irqreturn_t nanosic_wn8030_handler(int irq, void *data)
@@ -474,6 +566,8 @@ static irqreturn_t nanosic_wn8030_handler(int irq, void *data)
 	struct nanosic_wn8030 *nanosic = data;
 	u8 buf[XM_WN8030_I2C_READ];
 
+	/* After sleep pin deactivated, fw will trigger irq to notify about
+	   successful resume, second irq will tell about kb connection state */
 	if (nanosic->suspended) {
 		nanosic->suspended = false;
 		return IRQ_HANDLED;
@@ -500,11 +594,15 @@ static irqreturn_t nanosic_wn8030_handler(int irq, void *data)
 			hid_input_report(nanosic->hid_touchpad, HID_INPUT_REPORT,
 					 &buf[3], 21, 0);
 		break;
+	case 0x22:
 	case 0x23:
 		nanosic_wn8030_handle_vendor(nanosic, buf);
 		break;
 	}
 
+	if (buf[7] == 0x31 || buf[7] == 0x32 || buf[7] == 0x33 || buf[7] == 0x34 || buf[7] == 0x35)
+		print_hex_dump(KERN_INFO, "", DUMP_PREFIX_OFFSET, 16, 1, buf, sizeof(buf), false);
+
 	//print_hex_dump(KERN_INFO, "", DUMP_PREFIX_OFFSET, 16, 1, buf, sizeof(buf), false);
 
 	return IRQ_HANDLED;
@@ -719,6 +817,15 @@ static void nanosic_wn8030_power_off(struct nanosic_wn8030 *nanosic)
 	regulator_bulk_disable(ARRAY_SIZE(nanosic_wn8030_supply_names), nanosic->supplies);
 }
 
+static void nanosic_wn8030_wake_worker(struct work_struct *data)
+{
+	struct nanosic_wn8030 *nanosic =
+		container_of(data, struct nanosic_wn8030, wake_worker.work);
+	
+	nanosic_wn8030_set_kb_power(nanosic, true);
+	schedule_delayed_work(&nanosic->wake_worker, msecs_to_jiffies(12000));
+}
+
 static int nanosic_wn8030_probe(struct i2c_client *client)
 {
 	struct nanosic_wn8030 *nanosic;
@@ -769,6 +876,8 @@ static int nanosic_wn8030_probe(struct i2c_client *client)
 	dev_set_drvdata(nanosic->dev, nanosic);
 	i2c_set_clientdata(client, nanosic);
 
+	INIT_DELAYED_WORK(&nanosic->wake_worker, nanosic_wn8030_wake_worker);
+
 	ret = nanosic_wn8030_check_boot_id(nanosic);
 	if (ret)
 		goto err;
@@ -788,7 +897,7 @@ static int nanosic_wn8030_probe(struct i2c_client *client)
 
 	ret = devm_request_threaded_irq(&client->dev, client->irq,
 					NULL, nanosic_wn8030_handler,
-					IRQF_ONESHOT | IRQF_TRIGGER_RISING | IRQF_NO_SUSPEND,
+					IRQF_ONESHOT | IRQF_TRIGGER_RISING,
 					"nanosic_wn8030_irq", nanosic);
 	if (ret) {
 		ret = dev_err_probe(nanosic->dev, ret, "failed to request irq %d\n", client->irq);
@@ -806,6 +915,10 @@ static int nanosic_wn8030_probe(struct i2c_client *client)
 static void nanosic_wn8030_remove(struct i2c_client *client)
 {
 	struct nanosic_wn8030 *nanosic = i2c_get_clientdata(client);
+
+	disable_irq(nanosic->client->irq);
+	cancel_delayed_work_sync(&nanosic->wake_worker);
+
 	if (nanosic->hid_keyboard)
 		hid_destroy_device(nanosic->hid_keyboard);
 	if (nanosic->hid_touchpad)
@@ -817,8 +930,9 @@ static int nanosic_wn8030_resume(struct device *dev)
 {
 	struct nanosic_wn8030 *nanosic = dev_get_drvdata(dev);
 
-	gpiod_set_value_cansleep(nanosic->sleep_gpio, 0);
+	enable_irq(nanosic->client->irq);
 
+	gpiod_set_value_cansleep(nanosic->sleep_gpio, 0);
 	msleep(25);
 
 	if (nanosic->suspended) {
@@ -841,10 +955,11 @@ static int nanosic_wn8030_suspend(struct device *dev)
 {
 	struct nanosic_wn8030 *nanosic = dev_get_drvdata(dev);
 
-	gpiod_set_value_cansleep(nanosic->sleep_gpio, 1);
+	disable_irq(nanosic->client->irq);
+	cancel_delayed_work_sync(&nanosic->wake_worker);
 
+	gpiod_set_value_cansleep(nanosic->sleep_gpio, 1);
 	msleep(10);
-
 	nanosic->suspended = true;
 
 	return 0;
